tests/test_regularization.py

```diff
@@ -0,0 +1,436 @@
+def test_constant_param_scaler(lambda_, train_loss, val_loss):
+    scaler = updaters.ConstantParamScaler()
+    assert scaler(lambda_, train_loss, val_loss) == lambda_
+
+
+@pytest.fixture(
+    scope="module",
+    params=[
+        (0.5, (0.9, 1)),  # unlikely to fall inside the interval
+        (0.5, (0.01, 10)),  # likely to fall inside the interval
+    ],
+)
+def interval_param_scaler(request):
+    return updaters.IntervalParamScaler(*request.param)
+
+
+@pytest.mark.parametrize(
+    "lambda_",
+    [
+        10.0,
+        0.001,
+    ],
+)
+@pytest.mark.parametrize(
+    "train_loss",
+    [
+        th.tensor(100.0),
+        th.tensor(10.0),
+        th.tensor(0.1),
+        th.tensor(0.0),
+        100.0,
+        10.0,
+        0.1,
+        0.0,
+    ],
+)
+def test_interval_param_scaler(lambda_, train_loss, interval_param_scaler):
+    scaler = interval_param_scaler
+    tolerable_interval = scaler.tolerable_interval
+    scaling_factor = scaler.scaling_factor
+    eps = np.finfo(float).eps
+    if train_loss > eps:
+        # The loss is a non-zero scalar, so we can construct a validation loss for
+        # three different cases:
+
+        # case that the ratio between the validation loss and the training loss is
+        # above the tolerable interval
+        val_loss = train_loss * tolerable_interval[1] * 2
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 + scaling_factor)
+
+        # case that the ratio between the validation loss and the training loss is
+        # below the tolerable interval
+        val_loss = train_loss * tolerable_interval[0] / 2
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 - scaling_factor)
+
+        # case that the ratio between the validation loss and the training loss is
+        # within the tolerable interval
+        val_loss = train_loss * (tolerable_interval[0] + tolerable_interval[1]) / 2
+        assert scaler(lambda_, train_loss, val_loss) == lambda_
+    else:
+        # we have a zero loss. We try two cases. When the validation loss is zero,
+        # the ratio is undefined, so we should return the current lambda. When the
+        # validation loss is nonzero, the ratio is infinite, so we should see the lambda
+        # increase by the scaling factor.
+        # We try it for both a tensor and a float value.
+        val_loss = th.tensor(0.0)
+        assert scaler(lambda_, train_loss, val_loss) == lambda_
+        val_loss = 0.0
+        assert scaler(lambda_, train_loss, val_loss) == lambda_
+        val_loss = th.tensor(1.0)
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 + scaling_factor)
+        val_loss = 1.0
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 + scaling_factor)
+
+
+def test_interval_param_scaler_raises(interval_param_scaler):
+    scaler = interval_param_scaler
+    with pytest.raises(ValueError, match="val_loss and train_loss must be scalars"):
```

Comment:
Nit: because both train and val loss are illegal arguments, this error will trigger (and test will pass) if input validation checks *either* argument. Might want to split test cases to use legal argument for one not the other?

Not as relevant but it also seems odd to have `lambda_ = 0.0` the whole time as that's also illegal.

Reply:
Fair. Haha hope we don't get the AI optimizing for passing the checks only for the tensor `th.Tensor([3.0, 4.0])` ;)

Reply:
Actually, there's a cool library that I saw recently that tries to touch on that by having test values be automatically generated from assumptions (e.g. integer): https://hypothesis.readthedocs.io/en/latest/quickstart.html. And the test values are different every time.

Reply:
Oh lol, it seems like we already use it.